### PR TITLE
issue#101: Socks 4/5 Proxy handler in NetClient

### PR DESCRIPTION
**Pull request by alexlehm**
fixes https://github.com/vert-x3/issues/issues/101

add SOCKS proxy support to HttpClient and NetClient
changed proxy options into a ProxyObject class
add small socks4/5 proxy for testing
move proxyOptions property to ClientOptionsBase
add unit tests
add missing ctx.fireXXX in VertxHandler
move ProxyChannelProvider to its own package
add classnotfound check to NetClientImpl
add to documentation about proxy config

### CODE REVIEW

File: src/main/java/io/vertx/core/net/ProxyOptions.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  private String proxyHost;
+  private int proxyPort;
+  private String proxyUsername;
+  private String proxyPassword;
+  private ProxyType proxyType;
+
+  /**
+   * Default constructor.
+   */
+  public ProxyOptions() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/ProxyOptions.java]: default values needs to be constants, look at other data objects

File: src/main/java/io/vertx/core/net/ProxyOptions.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    proxyHost = "localhost";
+    proxyPort = 3128;
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/ProxyOptions.java]: in particular the port should be configured

File: src/main/java/io/vertx/core/net/ProxyOptions.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+   * @return a reference to this, so the API can be used fluently
+   */
+  public ProxyOptions setProxyType(ProxyType proxyType) {
+    Objects.requireNonNull(proxyType, "ProxyType may not be null");
+    this.proxyType = proxyType;
+    return this;
+  }
+
+  @Override
+  public ProxyOptions clone() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/ProxyOptions.java]: should use copy constructor, it's the common idiom for Vert.x data objects

File: src/main/java/io/vertx/core/net/ProxyType.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/**
+ * 
+ */
+package io.vertx.core.net;
+
+/**
+ * @author <a href="http://oss.lehmann.cx/">Alexander Lehmann</a>
+ *
+ */
+public enum ProxyType {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/ProxyType.java]: missing @VertxGen

File: src/main/java/io/vertx/core/net/ProxyType.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  HTTP, // HTTP CONNECT ssl proxy
+  SOCKS4, // SOCKS4/4a tcp proxy
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/ProxyType.java]: theses should be javadoc and not comments

**Comment by alexlehm** [comment]
superseded by #1439

**Pull request closed without merging**